packages/babel-parser/test/fixtures/experimental/pipeline-operator/proposal-fsharp-arrow-in-body-no-parens-indented-with-arg-parens/input.js

```diff
@@ -0,0 +1,3 @@
+(x) => x
+  |> (y) => y + 1
+  |> (z) => z * 2
```

Comment:
Maybe you could add a comment that this file is parsed as
```
(x) => (
    x |> (y) => (
             y + 1 |> (z) => z * 2
          )
)
```

Reply:
Shoot, I don't think that's what we actually want, because the variables from the previous pipeline function shouldn't be available in the next function's scope.

Reply:
So the underlying issue here is the arrow body's expression needs to stop parsing if it hits the pipeline operator token if it's nested within an arrow body expression already. I'm still poking around, but any thoughts / insights on where I might be able to implement this would be helpful.

Reply:
Just checking here ('cause I think I've hit this). `|>` should be left-associative and parsed like this, right?

```
(x) => ((x
  |> (y) => y + 1 )
  |> (z) => z * 2)
```

Reply:
Sorry for the long delay. How should this code be parsed?

Is it correct that `a = b |> c` is `a = (b |> c)` and `a |> () => b |> c` is `a |> (() => b) |> c`?

Is so, how whould this be parsed?
```js
a |> () => b = c |> d
```

a) `|>` is an "hard break", which escapes the precedence even if it is nested inside other expressions:
  ```js
  a |> (() => b = c) |> d
  ```
b) `|>` and `() =>`'s precedences are modified, but nothing else:
  ```js
  a |> () => b = (c |> d)
  ```

Reply:
It would be `a`–the mental shorthand is intended to be "pipeline operators terminate bodies of nested arrow functions".

Reply:
Ok, then you could add an `inFSharpPipelineDirectBody` flag to the state, and
1) Set it to `true` after `|>`
2) Set it to `false` inside `[({`/`})]`
3) If it is `true`, in `parseExprOp` return when you see `|>` instead of parsing it.

Another option is to pass a parameter (like `refShorthandDefaultPos`) down to the various parsing methods: it's not really Babel's style, but it is like you could probably write it in the spec.

I prefer the state flag.

Reply:
Thanks! Will take a run at that. Agreed on the state flag.

Reply:
@nicolo-ribaudo Updated mostly as suggested–I added tests for inside the `[` & `{` and I didn't have to change anything specific there for it to work as expected. Let me know if the AST looks off to you.